mlir-sycl/lib/Transforms/Inliner.cpp

```diff
@@ -0,0 +1,470 @@
+// Inliner
+//===----------------------------------------------------------------------===//
+
+LogicalResult Inliner::inlineSCC(Inliner &Inliner, CallGraphSCC &SCC,
+                                 MLIRContext *Ctx) {
+  Inliner::inlineCallsInSCC(Inliner, SCC);
+  return success();
+}
+
+/// Attempt to inline calls within the given scc. This function returns
+/// success if any calls were inlined, failure otherwise.
+void Inliner::inlineCallsInSCC(Inliner &Inliner, CallGraphSCC &SCC) {
+  for (CallGraphNode *SrcNode : SCC) {
+    if (SrcNode->isExternal())
+      continue;
+
+    Inliner::collectCallOps(*SrcNode->getCallableRegion(), SrcNode,
+                            Inliner.getCG(), Inliner.getSymbolTable(),
+                            Inliner.getCalls(),
+                            /*traverseNestedCGNodes=*/false);
+  }
+
+  if (Inliner.getCalls().empty())
+    return;
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "* Inliner: SCC: " << SCC << "\n";
+    llvm::dbgs() << "* Inliner: Initial calls in SCC are: {\n";
+    for (unsigned i = 0, e = Inliner.getCalls().size(); i < e; ++i)
+      llvm::dbgs() << "  " << i << ". " << Inliner.getCall(i).Call << ",\n";
+    llvm::dbgs() << "}\n";
+  });
+
+  // Try to inline each of the call operations. Don't cache the end iterator
+  // here as more calls may be added during inlining.
+  for (unsigned I = 0; I < Inliner.getCalls().size(); ++I) {
+    ResolvedCall It = Inliner.getCall(I);
+    CallOpInterface Call = It.Call;
+
+    bool DoInline = Inliner.shouldInline(It);
+    LLVM_DEBUG(llvm::dbgs() << ((DoInline) ? "* Inlining call: "
+                                           : "* Not inlining call: ")
+                            << I << ". " << Call << "\n";);
+    if (!DoInline)
+      continue;
+
+    Region *TgtRegion = It.TgtNode->getCallableRegion();
+    LogicalResult InlineRes = inlineCall(
+        Inliner, Call, cast<CallableOpInterface>(TgtRegion->getParentOp()),
+        TgtRegion, /*shouldCloneInlinedRegion=*/true);
+
+    if (failed(InlineRes)) {
+      LLVM_DEBUG(llvm::dbgs() << "** Failed to inline\n");
+      continue;
+    }
+
+    LLVM_DEBUG(llvm::dbgs() << "** Inline succeeded\n");
+
+    // then erase the call.
+    Call.erase();
+  }
+
+  Inliner.clear();
+}
+
+void Inliner::collectCallOps(iterator_range<Region::iterator> Blocks,
+                             CallGraphNode *SrcNode, CallGraph &CG,
+                             SymbolTableCollection &SymTable,
+                             SmallVectorImpl<ResolvedCall> &Calls,
+                             bool TraverseNestedCgNodes) {
+  SmallVector<std::pair<Block *, CallGraphNode *>, 8> WorkList;
+  auto AddToWorkList = [&](CallGraphNode *CGN,
+                           iterator_range<Region::iterator> Blocks) {
+    for (Block &B : Blocks)
+      WorkList.emplace_back(&B, CGN);
+  };
+
+  AddToWorkList(SrcNode, Blocks);
+
+  while (!WorkList.empty()) {
+    Block *B;
+    std::tie(B, SrcNode) = WorkList.pop_back_val();
```

Comment:
This reminds me we should enable C++ 17 in `mlir-sycl` as it is already enabled in LLVM.